### PR TITLE
[Salesforce] Delete operation

### PR DESCRIPTION
**Pull request by nick-Ag**
This PR implements the Salesforce delete operation for every action. The operation will lookup a record given some traits and if that record exists, call the delete endpoint using it's Id. This is being added to reach feature parity with the Salesforce (Classic) integration, which allows deletions.

## Testing

_Include any additional information about the testing you have completed to
ensure your changes behave as expected. For a speedy review, please check
any of the tasks you completed below during your testing._

- [x] Added [unit tests](https://github.com/segmentio/action-destinations/blob/main/docs/testing.md#local-end-to-end-testing) for new functionality
- [x] Tested end-to-end using the [local server](https://github.com/segmentio/action-destinations/blob/main/docs/testing.md#local-end-to-end-testing)
- [x] [Segmenters] Tested in the staging environment

The Delete operation being triggered:
Screenshot 2023-04-05 at 3 31 13 PM

The record being looked up for deletion:
Screenshot 2023-04-05 at 3 30 43 PM

The record is deleted:
Screenshot 2023-04-05 at 3 31 24 PM

### CODE REVIEW

File: packages/destination-actions/src/destinations/salesforce/__tests__/lead.test.ts [hunk 1]
```diff
@@ -70,6 +70,67 @@ describe('Salesforce', () => {
       )
     })
 
+    it('should delete a lead record given an Id', async () => {
```
Comment (brennan) [inline packages/destination-actions/src/destinations/salesforce/__tests__/lead.test.ts:73]: Can we add tests for all Salesforce object types, in addition to `Lead`? Thanks, @nick-Ag !

**Review by brennan — CHANGES_REQUESTED** [review]
Can we add tests for all Salesforce object types, in addition to Lead? Thanks, @nick-Ag !

**Comment by nick-Ag** [comment]
Hey @brennan , added those tests. Adding the `customObject` test revealed that the `customFields` field was required even for the delete operation, where it shouldn't be needed. So I made that field not required and only conditionally throw an error if it is not included when the operation is `create, update, upsert`. I stage tested and all looks good! Thanks for the testing suggestion.

Custom Fields required for custom action when operation is create:
Screenshot 2023-04-06 at 1 54 27 PM

No Custom Fields defined on this custom action, delete operation works fine:
Screenshot 2023-04-06 at 1 54 07 PM

File: packages/destination-actions/src/destinations/salesforce/__tests__/opportunity.test.ts [hunk 2]
```diff
@@ -131,6 +131,67 @@ describe('Salesforce', () => {
       )
     })
 
+    it('should delete an opprotunity record given an Id', async () => {
```
Comment (brennan) [inline packages/destination-actions/src/destinations/salesforce/__tests__/opportunity.test.ts:134]:
```suggestion
    it('should delete an opportunity record given an Id', async () => {
```

File: packages/destination-actions/src/destinations/salesforce/__tests__/opportunity.test.ts [hunk 3]
```diff
@@ -131,6 +131,67 @@ describe('Salesforce', () => {
...
+            Id: { '@path': '$.userId' }
+          }
+        }
+      })
+
+      expect(responses.length).toBe(1)
+      expect(responses[0].status).toBe(204)
+    })
+
+    it('should delete an opprotunity record given some lookup traits', async () => {
```
Comment (brennan) [inline packages/destination-actions/src/destinations/salesforce/__tests__/opportunity.test.ts:159]:
```suggestion
    it('should delete an opportunity record given some lookup traits', async () => {
```

File: packages/destination-actions/src/destinations/salesforce/customObject/index.ts [hunk 4]
```diff
@@ -41,6 +43,10 @@ const action: ActionDefinition<Settings, Payload> = {
     }
   },
   perform: async (request, { settings, payload }) => {
+    if (OPERATIONS_WITH_CUSTOM_FIELDS.includes(payload.operation) && !payload.customFields) {
+      throw new Error('Custom fields are required for this operation.')
```
Comment (brennan) [inline packages/destination-actions/src/destinations/salesforce/customObject/index.ts:47]: I think this should be an `IntegrationError`, similar to the other custom errors in Salesforce (Actions).

File: packages/destination-actions/src/destinations/salesforce/customObject/index.ts [hunk 5]
```diff
@@ -56,8 +62,16 @@ const action: ActionDefinition<Settings, Payload> = {
...
       return await sf.upsertRecord(payload, payload.customObjectName)
     }
+
+    if (payload.operation === 'delete') {
+      return await sf.deleteRecord(payload, payload.customObjectName)
+    }
   },
   performBatch: async (request, { settings, payload }) => {
+    if (OPERATIONS_WITH_CUSTOM_FIELDS.includes(payload[0].operation) && !payload[0].customFields) {
+      throw new Error('Custom fields are required for this operation.')
```
Comment (brennan) [inline packages/destination-actions/src/destinations/salesforce/customObject/index.ts:72]: Should be an `IntegrationError`, similar to the other custom errors in Salesforce (Actions).

**Review by brennan — APPROVED** [review]

**Pull request merged**